[feature] Implemented get text like translations #235

codesankalp opened the pull request:
Used ttag js to implement get text like translations.

Closes #235
Fixes #304 
Closes #308

nemesisdesign reviewed (changes requested):
Fix the conflict please.

nemesisdesign commented on .gitignore:
```diff
@@ -38,6 +38,9 @@ client/assets/*
 !client/assets/fonts/
+client/translations/
+i18n/*
+!i18n/en.po
```
shouldn't this be `!i18n/*.po`?
Otherwise we would include only english translations.

nemesisdesign commented on README.md:
```diff
@@ -270,41 +270,33 @@ Let us consider the following configuration for the header, footer and contact c
       # if organization supports mobile verification
-      - text:
-          en: "change phone number"
+      - text: "change phone number"
```
shouldn't we be able to remove this text from here?

nemesisdesign commented on README.md:
```diff
@@ -270,41 +270,33 @@ Let us consider the following configuration for the header, footer and contact c
         url: "/support"
-      - text:
-          en: "twitter"
+      - text: "twitter"
```
same here

nemesisdesign commented on README.md:
```diff
@@ -403,6 +389,37 @@ enable this feature, just create a new organization with the
 
 `Are you using OpenWISP Subscriptions to provide paid subscriptions for WiFi plans or identity verification via credit/debit card?`
 
+#### Translation
```
Translation > Translations?

nemesisdesign commented on README.md:
```diff
@@ -403,6 +389,37 @@ enable this feature, just create a new organization with the
 
+#### Translation
+
+1. All the translation is loaded at runtime from the JSON files that were made
```
translations are
were compiled instead of were made

nemesisdesign commented on README.md:
````diff
@@ -403,6 +389,37 @@ enable this feature, just create a new organization with the
+
+7. If there is more than one language in `i18n/` directory then update the organization configuration file by adding the support for that language like this:
+
+   ```
````
let's flag as yaml to get the colored syntax

nemesisdesign commented on package.json:
```diff
@@ -107,7 +110,9 @@
     "browser-test": "jest browser-test --runInBand",
     "test-inspect": "node inspect ./node_modules/.bin/jest --runInBand",
-    "stats": "STATS=true npx webpack-cli --profile --json --config config/webpack.config.js > compilation-stats.json && webpack-bundle-analyzer dist/stats.json"
+    "stats": "STATS=true npx webpack-cli --profile --json --config config/webpack.config.js > compilation-stats.json && webpack-bundle-analyzer dist/stats.json",
```
why npx?

codesankalp replied:
I removed it.

nemesisdesign reviewed (commented):
Terms & conditions and Privacy Policy: there's a problem here.
These will surely vary in each installation, so there's no point in adding these to the list of translatable texts.
Can you find a way to keep these as configuration please?

nemesisdesign reviewed (changes requested):
Another potential issue with the way language files are built and included:
I see that the language strings are included in the main chunk, but what happens if there are multiple languages, say 5 languages? Would all 5 langauges be included in the main chunk?

We should make sure each language is stored in its own file and then should be loaded only if the user is opening the app with that language.

nemesisdesign commented on i18n/en.po:
```diff
@@ -0,0 +1,660 @@
+
+#: client/components/contact-box/contact.js:22
+msgid "CONTACT_EMAIL"
+msgstr "support@openwisp.co"
```
:man_facepalming:  :facepalm:  :face_with_head_bandage:  :face_with_thermometer: 
these kind of things should always be configuration... Applies to any other organization specifc detail.

codesankalp replied:
I also don't want to define it here, but In configuration, I saw email is also defined for multiple languages i.e. old getText is used in the code for email value also.

codesankalp replied:
I am making email for single language only.

nemesisdesign commented on i18n/en.po:
```diff
@@ -0,0 +1,660 @@
+
+#: client/components/status/status.js:602
+msgid "DATE_LANGUAGE_LOCALE"
+msgstr "en-US"
```
how did this end up here? Does not look like a translatable string, it's just the language code, can you point out at what part of the code is generating this please?

codesankalp replied:
In status page for account start and stop time. Basically it translates the time to be in the format of that locale

nemesisdesign commented on internals/generators/organization/config.yml.hbs:
```diff
@@ -67,297 +64,133 @@ client:
       links:
-        - text:
-            en: "Sign In" #link text in english
+        - text: "Sign In" #link text in english
```
mmm...

What if a user wants to add a new link and translate it in multiple languages?
Can they do it somehow?

codesankalp replied:
Used previous getText

nemesisdesign commented on README.md:
````diff
@@ -403,6 +389,37 @@ enable this feature, just create a new organization with the
+     - text: "Spanish"
+       slug: "es"
+   ```
+
````
What about the possibility of changing/customizing translations?

codesankalp replied:
updated readme for this.

codesankalp commented on client/components/organization-wrapper/organization-wrapper.test.js:
```diff
@@ -214,7 +212,7 @@ describe("Test <OrganizationWrapper /> routes", () => {
   it("should display status if authenticated", () => {
     wrapper = mountComponent(props, ["/default/status"]);
-    expect(mapRoutes(wrapper)["/default/status"]).toBe(Suspense);
+    expect(mapRoutes(wrapper)["/default/status"]).toBe(undefined);
```
```suggestion
    expect(mapRoutes(wrapper)["/default/status"]).toBe(Suspense);
```
This is not giving Suspense when I tried. Do you know how to fix this?
Also, the component.find(Routes) returns empty but routes are present in that.
This is the only test where I faced problems in fixing.

nemesisdesign reviewed (changes requested):
Good progress @codesankalp :clap: :+1: !

There's a few minor conflicts to resolve due to the changes done in https://github.com/openwisp/openwisp-wifi-login-pages/pull/309, please take a look at the diff in the linked PR before merging the conflict.

See belo some issues I found during this round of testing.
This review is not exhaustive, I will be testing this work more in the next days while I work on the Italian translations.

### Be forgiving of not yet existing language translations

I had some translations defined in one of my testing confs, now the page does not load and this exception is logged in the JS console:

```
.*\.json$ namespace object:10 Uncaught (in promise) Error: Cannot find module './it.json'
    at .*\.json$ namespace object:10
```

Is it possible to change this part so that if the translation file is not found, the default language is used and a message is logged to console (eg: `console.error`)?

I have another question, what happens if a new string is added to the code of the application, but the translation is not yet available for every language? Will the code crash or will it display something anyway (and a message is logged to the console)? 

### Changing lang for an org changes it also for another org

Try having two orgs defined:

- default (leave default conf)
- second (add a lanaguage)

Now open the second org in a new tab and change the language to the other language you added, then go back to the tab with the default org and reload the page, what do you see?

I see this:

![Screenshot from 2021-07-12 21-59-44](https://user-images.githubusercontent.com/841044/125383106-843cb700-e35c-11eb-9f9c-b6162e282a21.png)

### Lang preferences

I see that you have addressed the second point of this issue: https://github.com/openwisp/openwisp-wifi-login-pages/issues/304, but you have not addressed the first, is that on purpose? We can do work on that after PR is complete if you prefer, I am just double checking with you.

codesankalp commented on client/components/logout/__snapshots__/logout.test.js.snap:
```diff
@@ -23,7 +23,39 @@ exports[`<Logout /> rendering should render correctly 1`] = `
+        className="main-column w-100"
+      >
+        <h2>
+          LOGOUT_CONTENT
```
I added a test for rendering with placeholder tags, so that in the future if placeholder tags are changed then the snapshot must fail, and updating `po` files are required. Same reason for other occurences also.

codesankalp commented on README.md:
```diff
@@ -420,6 +414,78 @@ enable this feature, just create a new organization with the
+   can be defined as the custom strings.
+4. Translation can also be generated for single organisation if `{orgSlug}_{language_code}.custom.po` is present in `i18n` Directory.
+5. Translation can be created by running the command:
+   `yarn create-translation {language_code} i18n/{file_name}.po`
```
for custom po files, all the msg-id is not required, so  I thought to add a script just to create a translation file with content headers.

nemesisdesign replied:
@codesankalp it sounds like we should make two separate commands, one for adding general translations and one for adding custom translations, then document these differently and explain the difference, how does this sound?

codesankalp commented:
> I have another question, what happens if a new string is added to the code of the application, but the translation is not yet available for every language? Will the code crash or will it display something anyway (and a message is logged to the console)?

It will display the translation placeholder on the webpage.

codesankalp commented on client/components/status/status.js:
```diff
@@ -650,6 +629,41 @@ export default class Status extends React.Component {
+      },
+      settings: {
+        active_session: t`ACCT_ACTIVE`,
+        date_language_locale: t`DATE_LANGUAGE_LOCALE`,
```
I will solve this in this.

codesankalp commented on README.md:
```diff
@@ -420,6 +414,78 @@ enable this feature, just create a new organization with the
+   `{language_code}.po`.
+
+6. Translations inside `i18n/` directory can be updated by using this command:
+   `yarn update-translation i18n/{file_name}.po client/`
```
Resolved

codesankalp commented on README.md:
````diff
@@ -420,6 +414,78 @@ enable this feature, just create a new organization with the
+       slug: "en"
+     - text: "Spanish"
+       slug: "es"
+   ```
````
Yes, it's still supported.

codesankalp replied:
Like it will use the default if browser returns none in `navigator.language`

codesankalp commented on package.json:
```diff
@@ -18,6 +18,7 @@
     "history": "^4.10.1",
     "js-yaml": "^3.14.0",
     "jsdom": "^16.4.0",
+    "lodash": "^4.17.21",
```
It's used in the build process only.

codesankalp commented:
> I see that you have addressed the second point of this issue: #304, but you have not addressed the first, is that on purpose? We can do work on that after PR is complete if you prefer, I am just double checking with you.

I implemented first part also, by using navigator. See the code in `load-translation.js`

```
if (useBrowserLang) lang = navigator.language || navigator.userLanguage;
```

codesankalp commented on client/utils/load-translation.js:
```diff
@@ -0,0 +1,63 @@
+  useBrowserLang = false,
+) => {
+  let lang = language;
+  if (useBrowserLang) lang = navigator.language || navigator.userLanguage;
```
Yes.

nemesisdesign replied:
@codesankalp I verified that having "en-US" as preferred language breaks this code

codesankalp replied:
Resolved

nemesisdesign commented on client/components/404/404.js:
```diff
@@ -4,39 +4,36 @@ import "./index.css";
           <div className="main-column">
             <div className="row owisp-404-row-1">
-              {page && page.heading ? getText(page.heading, language) : "Oops!"}
+              {page && page.heading ? t`404_HEADING` : "Oops!"}
```
I think we can remove these default values at the end right? Let's try to shorten the code a bit where the old code is not necessary anymore, keep in mind we have to start thinking in reducing the size of the build.

codesankalp replied:
I think No, because these default values are used on those URL where organization-slug is missing like `http://0.0.0.0:8080/error/` as setOrganization is not called. Also 404 page of organization and without organization is different.

nemesisdesign reviewed (changes requested):
Great progress @codesankalp.

I have done some improvements and pushed them to this branch, I also have the Italian translation file ready but I will send it in a separate PR when we are done with this.

While testing, I found quite a few non translated words in this file:
https://github.com/openwisp/openwisp-wifi-login-pages/blob/b6ef54a226fb2eb348e682041d73d6f7524f7e98/client/constants/index.js

See my comments below.

nemesisdesign reviewed (changes requested):
BTW, having custom translations breaks the test "load translation tests › should load". Is there anything you can do to prevent that?
It will be annoying for people like me who have to maintain different services and may want to keep the custom translations in the same directory without committing them.

codesankalp commented on README.md:
```diff
@@ -420,6 +414,100 @@ enable this feature, just create a new organization with the
+
+For extraction or updating translations in the `.po` file, Use the command:
+
+`POFILE=<path-to-po-file> yarn translations-update`
```
Because the first argument needs to be the path to po file and the second argument needs to be src files.

https://github.com/ttag-org/ttag-cli/blob/76b340e87b550193da03b9a3f88dbe88b648fc93/src/index.ts#L226-L236

codesankalp commented on internals/generators/organization/config.yml.hbs:
```diff
@@ -96,260 +93,107 @@ client:
-            en: "mobile phone number"
-          label:
-            en: "mobile phone number"
+          type: "tel"
```
Removed

codesankalp commented:
> BTW, having custom translations breaks the test "load translation tests › should load". Is there anything you can do to prevent that?
> It will be annoying for people like me who have to maintain different services and may want to keep the custom translations in the same directory without committing them.

Fixed this and added some more tests

nemesisdesign commented:
> Another thing, this will not work as expected because `plan.plan` needs to be msgID.

I did test this and it works. How can we work around the error?
I need to be able to translate that part which is defined in the DB but making the content in the DB translatable is really overkill and doing it here is going to be far faster to implement.

nemesisdesign commented on internals/generators/organization/config.yml.hbs:
```diff
@@ -249,6 +248,11 @@ client:
+          label:
+            en: "email"
+          placeholder:
+            en: "enter email"
```
I guess we can remove this since it's already present in the translations right?

codesankalp replied:
https://github.com/openwisp/openwisp-wifi-login-pages/pull/287#discussion_r671595861

nemesisdesign commented on org-configurations/default-configuration.yml:
```diff
@@ -181,6 +181,10 @@ client:
+          label:
+            en: "email"
+          placeholder:
+            en: "enter email"
```
same here

codesankalp replied:
I added it here because if the user wants a custom label then he can add it for the username field.

nemesisdesign commented on client/utils/load-translation.js:
```diff
@@ -0,0 +1,89 @@
+        lang = browserLang.substr(0, 2);
+        return false;
+      }
+      return true;
```
does this `true` and do the above `false` values returned here make any difference?
It's a bit confusing.

codesankalp replied:
It's just to break the iteration.

nemesisdesign replied:
then you should be able to simply use `return`, the last 2 return statements should not be needed, or is the linter complaining? Alternatively just use `else if` insead of the second `if` in which case you should be able to avoid the return completely and no ambiguity is left.

nemesisdesign commented on client/utils/load-translation.js:
```diff
@@ -0,0 +1,89 @@
+        );
+        await importTranslation(
+          orgSlug,
+          localStorage.getItem(`${orgSlug}-defaultLanguage`),
```
why should we get this from localStorage? Don't we have it in the conf?

codesankalp replied:
Ok, I will change this to not to use local storage.

nemesisdesign reviewed (changes requested):
Did another round of testing of the preferred browser language but it's not working as expected.

I have an org with two languages:

- Italian (default)
- English

My browser is set to prefer English.

The code does recognize this and the line `await importTranslation(orgSlug, language);` is executed without errors, however, the language on the page stays in Italian, the correct behavior should be to automatically switch to the preferred language of the user, which in this case is English.

codesankalp commented:
> Did another round of testing of the preferred browser language but it's not working as expected.
> 
> I have an org with two languages:
> 
> * Italian (default)
> * English
> 
> My browser is set to prefer English.
> 
> The code does recognize this and the line `await importTranslation(orgSlug, language);` is executed without errors, however, the language on the page stays in Italian, the correct behavior should be to automatically switch to the preferred language of the user, which in this case is English.

Have you tried this after clearing localStorage? I have tested the same scenario and English is being loaded in my case.

nemesisdesign commented on client/components/registration/registration.js:
```diff
@@ -259,6 +259,7 @@ export default class Registration extends React.Component {
   };
 
   getLabelText = (plan) => {
+    /* disable ttag */
```
nice :+1: 
is this to disable the warning?

codesankalp replied:
Yes

nemesisdesign reviewed (approved):
I found a way to improve/simplify the translations-update command and took advantage to update and improve the docs, please have a look at my last 2 commits.

The rest looks good, the main requirements are being met so I think we can merge this and move on, I will keep testing it since I will be adding translations for several languages, I will surely let you know if I find any bug.

codesankalp commented:
> I found a way to improve/simplify the translations-update command and took advantage to update and improve the docs, please have a look at my last 2 commits.

Yes, it looks good. I don't know about this command :sweat_smile:
 
> The rest looks good, the main requirements are being met so I think we can merge this and move on, I will keep testing it since I will be adding translations for several languages, I will surely let you know if I find any bug.

yes :smile:
